lib/conversion.c

```diff
@@ -0,0 +1,204 @@
+
+    return signature_format_err;
+}
+
+static void print_ssl_error(const char *failed_action) {
+    char errstr[256] = {0};
+    unsigned long errnum = ERR_get_error();
+
+    ERR_error_string_n(errnum, errstr, sizeof(errstr));
+    LOG_ERR("%s: %s", failed_action, errstr);
+}
+
+bool tpm2_convert_pubkey(TPM2B_PUBLIC *public, pubkey_format format, const char *path) {
+
+    if (format == pubkey_format_der || format == pubkey_format_pem) {
+        return tpm2_convert_pubkey_ssl(&((public->t).publicArea), format, path);
+    }
+    else if (format == pubkey_format_tss) {
+        // save raw as is
+        return files_save_bytes_to_file(path, (UINT8 *)public, sizeof(*public));
+    }
+
+    LOG_ERR("Unsupported public key output format.");
+    return false;
+}
+
+static bool tpm2_convert_pubkey_ssl(TPMT_PUBLIC *public, pubkey_format format, const char *path) {
+    bool ret = false;
+    FILE *fp = NULL;
+    RSA *ssl_rsa_key = NULL;
+
+    // need this before the first SSL call for getting human readable error
+    // strings in print_ssl_error()
+    ERR_load_crypto_strings();
+
+    if (public->type != TPM_ALG_RSA) {
+        LOG_ERR("Unsupported key type for requested output format. Only RSA is supported.");
+        goto error;
+    }
+
+    UINT32 exponent = (public->parameters).rsaDetail.exponent;
+
+    if (exponent == 0) {
+        exponent = RSA_DEFAULT_PUBLIC_EXPONENT;
+    }
+
+    // OpenSSL expects this in network byte order
+    exponent = tpm2_util_hton_32(exponent);
+
+    ssl_rsa_key = RSA_new();
+
+    if (!ssl_rsa_key) {
+        print_ssl_error("Failed to allocate OpenSSL RSA structure");
+        goto error;
+    }
+
+    ssl_rsa_key->e = BN_bin2bn((void*)&exponent, sizeof(exponent), NULL);
+    ssl_rsa_key->n = BN_bin2bn(
+            (public->unique).rsa.t.buffer,
+            (public->unique).rsa.t.size,
+            NULL);
+
+    if (!ssl_rsa_key->n || !ssl_rsa_key->e) {
+        print_ssl_error("Failed to convert data to SSL internal format");
+        goto error;
+    }
+
+    fp = fopen(path, "wb");
+
+    if (!fp) {
+        LOG_ERR("Failed to open public key output file '%s': %s",
+            path, strerror(errno));
+        goto error;
+    }
+
+    int ssl_res = 0;
+
+    switch(format) {
+    case pubkey_format_pem:
+        ssl_res = PEM_write_RSA_PUBKEY(fp, ssl_rsa_key);
+        break;
+    case pubkey_format_der:
+        ssl_res = i2d_RSA_PUBKEY_fp(fp, ssl_rsa_key);
+        break;
+    default:
+        break;
+    }
+
+    if (ssl_res <= 0) {
+        print_ssl_error("OpenSSL public key conversion failed");
+        goto error;
+    }
+
+    ret = true;
+
+error:
+    if (fp) {
+        fclose(fp);
+    }
+    if (ssl_rsa_key) {
+        RSA_free(ssl_rsa_key);
+    }
+    ERR_free_strings();
+
+    return ret;
+}
+
+bool tpm2_convert_signature(TPMT_SIGNATURE *signature, signature_format format, const char *path) {
+
+    if (format == signature_format_tss) {
+        /* TODO fix serialization */
```

Comment:
We actually have been discussing serialization issues on: https://github.com/01org/tpm2-tss/pull/533

The issue here is that this dumps all padding, so its output not portable across architectures.